[~] ansible role nvidia-docker

shellshock1953 opened the pull request:
Install Nvidia host kernel and docker driver.

vikram-overhaul committed:
first pass of azure compat

vikram-overhaul committed:
add option for iam role

vikram-overhaul committed:
make cloudinit work in azure

thecubed committed:
[azure-compat] remove dracut optimizations by amzn to allow booting in azure

thecubed committed:
[azure-compat] regenerate all initramfs instead of just booted kernel

thecubed committed:
[system-base] fix cloud detection logic

shellshock1953 committed:
ansible role nvidia-docker

shellshock1953 committed:
providers/amazon/images/ivy-base/provision.yml - role nvidia-docker

shellshock1953 committed:
providers/amazon/images/ivy-base/provision.yml - role nvidia-docker

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,72 @@
+    name: gcc10.x86_64
+    state: present
+
+- name: "[nvidia-docker] Save previos GCC binary"
```
I'm a bit worried about messing with GCC defaults. Can we provide an environment variable to the NVIDIA driver installer to tell it to use `/bin/gcc10-gcc` as it's compiler? Usually you can do something like
```
GCC=/bin/gcc10-gcc /root/NVIDIA-installer.run
```
to cause it to pick that as the preferred version of GCC, that way we don't have to change system defaults.

We definitely have a weird case here since the kernel is built with new GCC, but the rest of userland is built with old GCC, so we don't want to replace the default GCC - just use the new one for building kernel-related things

shellshock1953 replied:
Used alternatives for this one + always switch to gcc7 as the last task in this role.

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,72 @@
+  debug: 
+    msg: "{{ nvidia_install_log.stdout_lines }}"
+
+- name: "[nvidia-docker] nvidia-docker repo"
```
Yep, let's use yum_repository instead of shell scripting here since this isn't super idempotent

shellshock1953 replied:
done!

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,72 @@
+    name: docker
+    state: restarted
+
+- name: "[nvidia-docker] validate host driver"
```
Fantastic idea to use validation in the build process!
An issue, though, is that we don't actually build the AMI on a host with a GPU. We build AMIs in a chroot, so they aren't even running the right kernel at build time.

This brings up a bigger issue... installing the NVIDIA drivers inside the chroot will cause it to try to build the kernel module for the kernel that the AMI builder is running - which is likely something old.

Can the NVIDIA installer be told to build for a specific kernel version? Or does the installer use DKMS?

shellshock1953 replied:
Nvidia driver refuses to install without GPU
```
You do not appear to have an NVIDIA GPU supported by the 510.54 NVIDIA Linux graphics driver installed in this system.
```
Checked possible options/args/etc to skip validation, but no luck here...
Can we make option to build nvidia in ec2 instead of chroot?

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,72 @@
+  # TODO: rewrite this to yum_repositiry
+  shell: "curl -s -L https://nvidia.github.io/nvidia-docker/$(. /etc/os-release;echo $ID$VERSION_ID)/nvidia-docker.repo | sudo tee /etc/yum.repos.d/nvidia-docker.repo"
+
+- name: "[nvidia-docker] Install nvidia-docker2"
```
Excellent!

Question: I know that kubernetes is deprecating dockershim support in 1.23. Does NVIDIA have an alternative that works with `containerd` or something?

...or is this a 'punt for later' thing? I'm alright if that's what we want to do.

shellshock1953 replied:
Seems they have support for [containerD](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#id6), still need to be tested first.

missingcharacter committed:
Adding support for enabling datadog's network monitoring (#35)

* Adding support for enabling datadog's network monitoring

* enable_datadog_cloud_security and enable_datadog_network_monitoring now use
install command instead of sudo cp

missingcharacter committed:
Bumping datadog-agent version to `7.33.1`

version `7.32.4` does not have `NPM DNS domain collection` since
it was introduced on version `7.33.0`
https://github.com/DataDog/datadog-agent/blob/main/CHANGELOG.rst#new-features

shellshock1953 commented on providers/amazon/images/ivy-base/provision.yml:
```diff
@@ -5,4 +5,5 @@
-    - consul
\ No newline at end of file
+    - consul
+    - nvidia-docker
```
done!

shellshock1953 commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,72 @@
+- name: "[nvidia-docker] Install GCCv10"
```
Nuked!

shellshock1953 committed:
nvidia-docker - on amazon/images/ivy-kubernetes

shellshock1953 committed:
nvidia-docker - on amazon/images/ivy-kubernetes

shellshock1953 committed:
nvidia-docker - role refactor

missingcharacter committed:
k8s-controllers: Bumping e2d version to 0.4.14-nxt+6 (#38)

This version looks for `DeviceIndex: 0` and no longer looks for NetworkInterfaces[0]

missingcharacter committed:
Bumping base Amazon Linux 2 ami to amzn2-ami-hvm-2.0.20220207.1-x86_64-gp2 (#39)

missingcharacter committed:
Accelerate checking k8s-agents and k8s-controllers setup via 1 script

missingcharacter committed:
[k8s-check.sh] EXPLAIN defaults to no

missingcharacter committed:
[k8s-check.sh] run not ran

missingcharacter committed:
[k8s-check.sh] check_systemctl_status now confirms status

missingcharacter committed:
[k8s-check.sh] now prints usage when unknown option is passed
[bash_lib/k8s.sh] New function `get_k8s_node_name` gets the name of the node from instance id
[bash_lib/k8s.sh] New function `get_k8s_node_status` gets node status
[bash_lib/k8s.sh] check_systemctl_status now uses systemctl is-active
[bash_lib/k8s.sh] k8s_controller_checks changes:
- Now list etcd members and stores them in an array
- Checks there are 3 members
- Checks 3 members are healthy
- Checks there is 1 leader
[bash_lib/k8s.sh] k8s_checks changes:
- Checks instance is part of k8s cluster using get_k8s_node_name
- Checks instance is `Ready` using get_k8s_node_status
- `datadog agent` times out after 2 seconds
- NAT hairpin is only checked when using CNI `aws-vpc-cni`

missingcharacter committed:
[bash_lib/k8s.sh] shellcheck is now happier

missingcharacter committed:
[WIP] Running shellcheck on files

missingcharacter committed:
Adding pre-commit hooks

missingcharacter committed:
[*.sh] shellcheck now passes

missingcharacter committed:
[bash_lib/k8s.sh] We no longer check for CSR
we now check for datadog-agent status, features and no failed checks

missingcharacter committed:
[k8s-checks.sh] Removing duplicate hashbang

missingcharacter committed:
Creating dnsmasq ansible role
amazon linux and dnsmasq https://aws.amazon.com/premiumsupport/knowledge-center/dns-resolution-failures-ec2-linux/
Retiring ivy-kubernetes 1.20
Bumping base amazon linux 2 to amzn2-ami-hvm-2.0.20220218.1-x86_64-gp2

missingcharacter committed:
Disabling SC1091 on dnsmasq/files/opt/ivy/set-dnsmasq-dhcp.sh

missingcharacter committed:
[bash_lib/k8s.sh] function check_dd_features now only checks for apm, logs and networks

missingcharacter committed:
Shellcheck is now happy with all files in repo

missingcharacter committed:
pipe to grep and awk, instead of redirecting command output

missingcharacter committed:
[dnsmasq] Ran `pre-commit run --all-files`

missingcharacter committed:
[shellcheck] Rebased from `master` branch and ran `pre-commit run --all-files`

missingcharacter committed:
[build.sh] get_packer_vars checks for empty line

missingcharacter committed:
[set-dnsmasq-dhcp.sh] DNS now gets the last domain-name-servers
[bash_lib/k8s.sh] [generate_pki] Replaced EXT with OPENSSL_ARGS array

missingcharacter committed:
[roles/dnsmasq] On Amazon Linux we now enable dnsmasq 2.85 repo
[build.sh] now supports long arguments, region and packer vars are no longer specified via comma separated values but by passing the argument multiple times
[bash_lib/k8s.sh] Rollback OPENSSL_ARGS and brought EXT with # shellcheck disable=SC2086

missingcharacter committed:
Updating base image to amzn2-ami-hvm-2.0.20220218.3-x86_64-gp2

missingcharacter committed:
[roles/kubernetes] ivy scripts are now copied via a loop

missingcharacter committed:
[bash_lib/k8s.sh] Checking for etcd leaders now trims whitespace after cut
before it was trimming after echo

missingcharacter committed:
[k8s-check.sh] Now prints a message when all checks exited without errors

missingcharacter committed:
[system-base] Install yum-plugin-versionlock and use it to lock pritunl and datadog-agent
[bash_functions.sh] function setup_docker_storage now enables or restarts docker depending on whether
docker is already running or not

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,73 @@
+    dest: "/root/NVIDIA-{{ nvidia_driver_version }}.run"
+
+- name: "Install nvidia-{{ nvidia_driver_version }} driver"
+  shell: "sh /root/NVIDIA-{{ nvidia_driver_version }}.run --silent"
```
It looks like the NVIDIA driver takes some advanced options that might help us when running in a chroot with a different kernel:

```
  -k KERNEL-NAME, --kernel-name=KERNEL-NAME
      Build and install the NVIDIA kernel module for the non-running kernel specified by KERNEL-NAME (KERNEL-NAME should be the output of `uname -r` when the target kernel is actually running).  This option implies '--no-precompiled-interface'.  If the options '--kernel-install-path' and '--kernel-source-path' are not given, then they will be inferred from KERNEL-NAME; eg: '/lib/modules/KERNEL-NAME/kernel/drivers/video/' and '/lib/modules/KERNEL-NAME/build/', respectively.
  --dkms
      nvidia-installer can optionally register the NVIDIA kernel module sources, if installed, with DKMS, then build and install a kernel module using the DKMS-registered sources.  This will allow the DKMS infrastructure to automatically build a new kernel module when changing kernels.  During installation, if DKMS is detected, nvidia-installer will ask the user if they wish to register the module with DKMS; the default response is 'no'.  This option will bypass the detection of DKMS, and cause the installer to attempt a DKMS-based installation regardless of whether DKMS is present.
```

The `--dkms` option will ensure that the module gets rebuilt after a new kernel is installed, which is useful since our systems auto-upgrade.

We can use the `-k` option to tell it to use the kernel that was installed by kernel-ml.
You can find the version of the kernel the system will use on next boot by running `grubby --default-kernel`, which should output something like:
```
/boot/vmlinuz-5.16.13-1.el7.elrepo.x86_64
```

The installer also has some options that would be useful to limit the cruft that gets installed:

```
  --no-wine-files
      Do not install any of the Wine driver files.
```

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,73 @@
+    - nvidia-container-runtime
+    - nvidia-docker2
+
+- name: "Restart docker"
```
This might not work in the chroot environment, since docker isn't running inside there anyway

thecubed commented on roles/nvidia-docker/tasks/main.yml:
```diff
@@ -0,0 +1,73 @@
+  systemd:
+    name: docker
+    state: restarted
+
```
At the end of this, can we delete the nvidia driver install file since it's rather large?

thecubed commented on providers/amazon/images/ivy-kubernetes/provision.yml:
```diff
@@ -7,7 +7,8 @@
     - system-base
     - docker
     - kubernetes
+    - nvidia-docker
```
We recently removed the `ivy-kubernetes` image, and moved to versioned roles. You should add this line to the `ivy-kubernetes-1.21` image now. Sorry for the confusion that I caused with my prior comment.

thecubed commented on roles/nvidia-docker/defaults/main.yml:
```diff
@@ -0,0 +1,2 @@
+nvidia_driver_version: 510.54
+nvidia_driver_validate: true
```
Can we default this to false, and we can enable it explicitly if we're running outside a chroot?

missingcharacter committed:
[WIP] [build.sh] works after binaries are present at roles/system-base/files/usr/local/bin/
[packer] missing python2 rpm bindings

missingcharacter committed:
ivy-base on fedora:35 works!

missingcharacter committed:
Created Fedora specific tasks for docker ansible role

missingcharacter committed:
[roles/system-base] Moved kernel-devel-matched to Fedora.yml

missingcharacter committed:
Install docker on fedora

missingcharacter committed:
Updating amazon provider ansible to version 5.4.0

missingcharacter committed:
Adding pre-commit github action

missingcharacter committed:
Updating markdownlint-cli pre-commit hook to https

missingcharacter committed:
[github-action] [pre-commit] Install shellcheck from homebrew

missingcharacter committed:
[GitHub Action] Updating checkout and setup-python actions to v3

missingcharacter committed:
Adding Github Actions validation and updating to ansible==5.6.0 (#46)

* Updating to ansible==5.6.0
Creating Github Action that will run docker build to test ansible and system-base

* [Github Action] run tests now sets TERM to xterm-256color

* [scripts/binaries/download_binaries.sh] chmod now uses sudo if user is NOT root

shellshock1953 committed:
roles/nvidia-docker - -k --dkms
